### ISSUE TITLE
Terminology: "teleprompter" and "compile"

### ISSUE THREAD

**Issue by jflam** [stanfordnlp/dspy#151]
Thanks, @okhat for giving a talk with help from Jerry and the folks over at LlamaIndex.

https://www.youtube.com/watch?v=POBcYr0sbcg

I really like the analogy to pytorch. But every time you talked about teleprompter in the talk, you also said "optimizer" and a few times that you said "compile" you also said "train".

I was also confused by these terms the first time I encountered them in DSPy. While this would be a breaking change to rename these terms in the library, I also think it will help new users and grow adoption of it if you double-down on the pytorch analogy in the code and the examples. 

We would be able to say "I am training my program using the BootstrapFewShotWithRandomSearch optimizer" and that would be much clearer than saying "I am compiling my program using the BootStrapFewShotWithRandomSearch teleprompter".

Thoughts? Thanks for making this excellent library!

**Comment by darinkishore** [comment]
I see your point, but disagree with using the word training. They're similar enough, but if you're using an LM like 3.5-turbo, which I think is one of the most popular use-cases, you're optimizing a prompt/data input format based on a certain metric. This is similar to training, but isn't quite actually fine-tuning the weights of the model based on a huge set of training data. 

The bootstrapfewshot w/random search teleprompter (to the best of my knowledge, please correct me if I am wrong) is going to find bunch of different examples and intermediate steps for your task, showcase that to the LM, and have it get the best outputs for the metric, refining the examples and intermediate steps based on metric performance. I'd say this is vastly different from a "training" pipeline—Optimizing and compiling seems like the best terminology to me. 

However, there is merit to say you're training the LM to do a task. It's definitely a simpler way of putting it. However, the how's of training are totally different from actual training in this case, so that opens up the road to a lot of confusion further down the line, especially if you're *actually* training with DSPy, like a T5-5b/LLama model or the like. If I had to guess, that's most likely why the training terminology wasn't used for teleprompters/compiling pipelines. 

Please tell me what you think! Huge fan of the library as well.

**Comment by jflam** [comment]
Thanks. A few of us were discussing the paper yesterday at work, and we agree that there isn't a great fit for existing terms for what DSPy aims to do.

If we think about compiling as being the process of taking something and transforming it to something else before executing, then it does describe what DSPy does. There's also an optimization step which happens during compilation, and we started riffing about PGO - profile guided optimization as one of the cases where in addition to the program, there is also data fed into the compiler to produce the final result.

So perhaps "I am compiling my program using the BootStrapFewShotWithRandomSearch optimizer"? So the idea is that the user gets to select different optimizers, kind of like what LLVM does with their back-end: https://llvm.org/docs/CodeGenerator.html

**Comment by darinkishore** [comment]
Agreed on no great fit!

As far as compilization and optimization, they both intuitively make sense for what dspy is trying to do. 

But, I think the words fall apart when you do something like 2x bootstrap, though, right? You should never have to double compile to produce same output, but here, double compile gives significantly better output. Same with optimize. 

See:

1: Compile then optimize -> Good program

2. Good program -> compile then optimize -> better program


So logically, the words make sense, but there's something off about it. Why would a program ever need to be compiled again? But also, compilation always takes smth unique and returns smth unique, so???

Hope it's clear what I'm getting at, because this is hard to put into words.

**Comment by jflam** [comment]
Yeah, this is a complex topic yet words are all we have to describe it :) 

There is precedent in compilation for compiling more than once. The JVM pioneered deoptimization where JIT-compiled code is reverted to interpretation to gather more data before (potentially) JIT compiling again. But that requires some runtime data which is not happening here. But that also suggests that this technique could be applied against a running system using feedback from users (or perhaps even another model) to trigger a compilation with new data when some threshold is met.

Certainly it's early days in all of this. Thanks for indulging me on this idea.

**Comment by okhat** [comment]
I'm coming around to considering the term "optimizer" instead of teleprompter.

But we'll definitely want to keep "compile". See https://twitter.com/lateinteraction/status/1716967956050247798
